macros/ui/niceTables.pl

```diff
@@ -3,835 +3,1201 @@ =head1 niceTables.pl
+	$html = prefix($html, $htmlcaption) unless $tableOpts->{LaYoUt};
+
+	if ($tableOpts->{LaYoUt}) {
+		$css .= css('display',         'table');
+		$css .= css('border-collapse', 'collapse');
+		$html = tag($html, 'div', { style => $css });
+	} else {
+		$html = tag($html, 'table', { style => $css });
+	}
+
+	# PTX
+	my $ptx     = $rows;
+	my $ptxleft = getPTXthickness($alignment[0]->{left});
+	my $ptxtop  = ($tableOpts->{horizontalrules}) ? 'major' : '';
+	$ptxtop = getPTXthickness($top) if $top;
+	my $ptxwidth;
+	my $ptxmargins;
+
+	if ($hasX) {
+		$ptxwidth = $tableOpts->{Xratio} * 100;
+		my $leftmargin  = ($tableOpts->{center}) ? (100 - $ptxwidth) / 2 : 0;
+		my $rightmargin = 100 - $ptxwidth - $leftmargin;
+		$ptxmargins = "${leftmargin}% ${rightmargin}%";
+		$ptxwidth .= '%';
+	} elsif (!$tableOpts->{center}) {
+		$ptxwidth   = '100%';
+		$ptxmargins = '0% 0%';
+	}
+	my $ptxbottom = ($tableOpts->{horizontalrules}) ? 'minor' : '';
+	if ($tableOpts->{LaYoUt}) {
+		$ptx = tag(
+			$ptx,
+			'sbsgroup',
+			{
+				width   => $ptxwidth,
+				margins => $ptxmargins,
+			}
+		);
+	} elsif (!$tableOpts->{LaYoUt}) {
+		$ptx = prefix($rows, $cols);
+		$ptx = tag(
+			$ptx,
+			'tabular',
+			{
+				valign  => ($tableOpts->{valign} != 'middle') ? $tableOpts->{valign} : '',
+				width   => $ptxwidth,
+				margins => $ptxmargins,
+				left    => $ptxleft,
+				top     => $ptxtop,
+				bottom  => $ptxbottom
+			}
 		);
-	my $center = $options{center};
-	if ($center != 0) { $tablecss .= 'text-align:center; margin:0 auto; ' }
-
-	# shortcuts introduced late
-	if (defined $options{align}) { $texalignment = $options{align} }
-	if ($options{rowheaders} == 1) {
-		for my $i (0 .. $#{$dataref}) {
-			${ $dataref->[$i][0] }{header} = 'RH' if (uc(${ $dataref->[$i][0] }{header}) eq '');
+	}
+
+	# We fake a caption as a following tabular
+	my $ptxcaption = '';
+	if ($tableOpts->{caption}) {
+		$ptxcaption = $tableOpts->{caption};
+		$ptxcaption = tag($ptxcaption, 'cell');
+		$ptxcaption = tag($ptxcaption, 'row');
+		my $ptxcapwidth = '';
+		if ($hasX) {
+			$ptxcapwidth = $tableOpts->{Xratio} * 100 . '%';
+		} else {
+			$ptxcapwidth = '50%';
 		}
+		$ptxcapcol  = tag('', 'col', { width => $ptxcapwidth });
+		$ptxcaption = prefix($ptxcaption, $ptxcapcol);
+		$ptxcaption = tag($ptxcaption, 'tabular', { width => $ptxwidth, margins => $ptxmargins });
 	}
+	$ptx = suffix($ptx, $ptxcaption);
+
+	return main::MODES(
+		TeX  => $tex,
+		HTML => $html,
+		PTX  => $ptx,
+	);
+
+}
+
+sub Cols {
+	my ($alignment, $tableOpts, $optsArray) = @_;
+	my $columnscss = $tableOpts->{columnscss};
+	my @html;
+	my @ptx;
+	my @alignment = @$alignment;
+	my $leftend   = shift(@alignment);
+
+	for my $i (0 .. $#alignment) {
+		my $align = $alignment[$i];
+
+		# determine if this column has paragraph cells
+		my $width = '';
+		for my $y (@$optsArray) {
+			if ($y->[$i]->{halign} =~ /^p\{([^}]*?)\}/) {
```

Comment:
`$y->[$i]->{halign}` can be undefined, and if so gives warnings about the use of an uninitialized value in pattern match.